incr: cache dwarf objects in work products

davidtwco opened the pull request:
Cache DWARF objects alongside object files in work products when those exist so that DWARF object files are available for thorin in packed mode in incremental scenarios.

r? @michaelwoerister

bjorn3 commented on compiler/rustc_incremental/src/persist/work_product.rs:
```diff
@@ -13,43 +13,58 @@ use std::path::Path;
     cgu_name: &str,
-    path: &Path,
+    object_file_path: &Path,
+    dwarf_object_file_path: Option<&Path>,
```
I did prefer an interface that allows arbitrary files. For cg_clif I want to be able to cache compiled inline assembly under the same workproduct so that I don't have to do a partial link of the main object file and the compiled inline assembly. Partial links are not supported on Windows AFAIK and are broken on macOS. This is one of the main blockers for fully functioning inline assembly support in cg_clif.

davidtwco replied:
I decided to add to `WorkProduct` because it looked like the rest of the compiler was assuming a single `WorkProduct` per CGU.

I could change this pull request to just store a `Vec<String>` in the `WorkProduct`, or I could make it easier for multiple `WorkProduct`s to be associated with a single CGU. In either case, we'd probably want some way of working out what a `WorkProduct` has without just working it out from filenames? Or I could leave it as-is in the pull request and this mechanism can be made more generic in a follow-up (which I'd be happy to do).

bjorn3 replied:
> I could change this pull request to just store a Vec<String> in the WorkProduct

This or having `HashMap<String, String>` would be my preferred interface.

> or I could make it easier for multiple WorkProducts to be associated with a single CGU.

This makes less sense to me as all compiled results of a cgu are changed together and a `WorkProduct` is the smallest unit of incremental compilation for codegen.

davidtwco replied:
I've updated this pull request so that `WorkProduct` contains a `HashMap<String, String>`

bjorn3 replied:
https://github.com/bjorn3/rustc_codegen_cranelift/commit/e45f6000a0bd46d4b7580db59c86f3d30adbc270 implemented what I was talking about above.

michaelwoerister commented:
Looks good, @davidtwco! Now we just need a regression test. Compiling twice with `-Zsplit-dwarf-kind=split -Csplit-debuginfo=packed` and no changes in between triggered bug for me.

davidtwco commented:
> Looks good, @davidtwco! Now we just need a regression test. Compiling twice with `-Zsplit-dwarf-kind=split -Csplit-debuginfo=packed` and no changes in between triggered bug for me.

I'm struggling to reproduce the issue with these instructions - using a simple case and compiling twice with those flags doesn't cause an issue for me.

davidtwco commented:
> Looks good, @davidtwco! Now we just need a regression test. Compiling twice with `-Zsplit-dwarf-kind=split -Csplit-debuginfo=packed` and no changes in between triggered bug for me.

Okay, added a test now. I lost some time to a missing `-g` flag in the test I wrote, which was a little frustrating, and then it turned out that `link_dwarf_object` wasn't calling `abort_if_errors` so tests wouldn't fail when `thorin` had.

I also noticed that we weren't deleting all the DWARF object temporaries in some cases, so I fixed that - we weren't deleting `module.dwarf_object` for allocator or metadata modules.
